src/objective/rank_obj.cu

```diff
@@ -0,0 +1,651 @@
+    // to store them in dsorted_labels_
+    // doriginal_pos_:      0 2 1 3 4 6 7 5 8 9  (stays the same)
+    // dsorted_labels_:     1 1 0 2 1 3 3 1 4 4  (from unsorted dlabels - dlabels)
+    thrust::gather(doriginal_pos_.begin(), doriginal_pos_.end(),
+                   dlabels.tcbegin(), dsorted_labels_.begin());
+  }
+
+  ~SortedLabelList() {
+    dh::safe_cuda(cudaSetDevice(device_id_));
+  }
+
+  // This kernel can only run *after* the kernel in sort is completed, as they
+  // use the default stream
+  void ComputeGradients(const bst_float *dpreds,
+                        GradientPair *out_gpair,
+                        const HostDeviceVector<bst_float> &weights,
+                        float weight_normalization_factor) {
+    // Group info on device
+    const unsigned *dgroups = dgroups_.data().get();
+    int ngroups = dgroups_.size();
+
+    int total_items = group_segments_.size();
+    int niter = param_.num_pairsample * total_items;
+
+    float fix_list_weight = param_.fix_list_weight;
+
+    const int *original_pos = doriginal_pos_.data().get();
+
+    size_t num_weights = weights.Size();
+    auto dweights = num_weights ? weights.ConstDevicePointer() : nullptr;
+
+    const bst_float *sorted_labels = dsorted_labels_.data().get();
+
+    // For each instance in the group, compute the gradient pair concurrently
+    dh::LaunchN(device_id_, niter, nullptr, [=] __device__(size_t idx) {
+      // First, determine the group 'idx' belongs to
+      unsigned item_idx = idx % total_items;
+      int group_idx = dh::UpperBound(dgroups, ngroups, item_idx);
+      // Span of this group within the larger labels/predictions sorted tuple
+      int group_begin = dgroups[group_idx - 1];
+      int group_end = dgroups[group_idx];
+      int total_group_items = group_end - group_begin;
+
+      // Are the labels diverse enough? If they are all the same, then there is nothing to pick
+      // from another group - bail sooner
+      if (sorted_labels[group_begin] == sorted_labels[group_end - 1]) return;
+
+      // Find the number of labels less than and greater than the current label
+      // at the sorted index position item_idx
+      int nleft  = LowerBound(sorted_labels + group_begin, total_group_items,
+                              sorted_labels[item_idx]);
+      int nright = UpperBound(sorted_labels + group_begin, total_group_items,
+                              sorted_labels[item_idx]);
+      nleft  = (nleft  == -1) ? 0 : nleft + 1;
+      nright = (nright == -1) ? 0 : total_group_items - nright;
+
+      // Create a minstd_rand object to act as our source of randomness
+      thrust::minstd_rand rng;
+      rng.discard(idx);
+      // Create a uniform_int_distribution to produce a sample from outside of the
+      // present label group
+      thrust::uniform_int_distribution<int> dist(0, nleft + nright - 1);
+
+      int sample = dist(rng);
```

Comment:
What if some labels are the same?  Say within a group of 4 elements, 2 of them are the same?

Reply:
while processing each label, it will pick the other one. if it is the left most pair, it will use one from the right most pair and vice-versa. the test demonstrates this.

Reply:
@sriramch Thanks, will take a closer look into this tomorrow. ;-)